samples/secret-rotation/index.js

```diff
@@ -0,0 +1,58 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+'use strict';
+
+// [START secretmanager_rotate]
+const {SecretManagerServiceClient} = require('@google-cloud/secret-manager');
+const crypto = require('crypto');
+
+// Instantiates a secret manager client.
+const client = new SecretManagerServiceClient();
+
+// The secret to rotate.
+const secret_name = process.env.SECRET_NAME;
+
+// Process a rotate notification to add a new SecretVersion to an existing
+// Secret.
+exports.rotate = async (event, context) => {
+  // Print the message data
+  console.log(`Received: ${event}`);
+  console.log(`Context: ${context}`);
+
+  // If the same topic is use for multiple secret rotators then we must check
+  // that the message is for the same secret that this function should rotate.
+  const request = Buffer.from(event.data, 'base64').toString();
+  console.log(`Requested rotation: ${request}`);
+  if (request !== secret_name) {
```

Comment:
So that samples can be embedded as snippets on cloud.google.com, we should follow the following sample guidelines for structure:

```js
function main(arg1 = 'default_value_one', arg2 = 1234) {
  // [START your_region_tag]
  /**
   * TODO(developer): Uncomment these variables before running the sample.
   */
  // const arg1 = 'default_value_one';
  // const arg2 = 1234;

  // Imports the Google Cloud Some API library
  const {SomeApiClient} = require('@google-cloud/some-api');

  // Instantiates a client
  const someApiClient = new SomeApiClient();

  async function doSomething() {
    ...
  }

  doSomething();
  // [END your_region_tag]
}

main(...process.argv.slice(2));
```

This is because top level await is not yet supported by our libraries.

Reply:
I'm familiar with the 'main' syntax like that, but I'm trying to make a sample that is a runnable cloud function like:

https://cloud.google.com/functions/docs/tutorials/pubsub
https://cloud.google.com/functions/docs/tutorials/use-cloud-spanner#prepare_the_application

And other samples under https://github.com/GoogleCloudPlatform/nodejs-docs-samples/tree/master/functions

That the sample includes the cloud function specific code is important to the sample. Does it still make sense to wrap this in a `main` like that? It seems like that would make it more difficult to work with cloud functions or the `functions-framework` test harness?

Reply:
If we want it embeddable in a cloud function, you could instead write like this:

```js
function main(arg1 = 'default_value_one', arg2 = 1234) {
  // [START your_region_tag]
  /**
   * TODO(developer): Uncomment these variables before running the sample.
   */
  // const arg1 = 'default_value_one';
  // const arg2 = 1234;

  // Imports the Google Cloud Some API library
  const {SomeApiClient} = require('@google-cloud/some-api');

  // Instantiates a client
  const someApiClient = new SomeApiClient();

  async function handler(req, res) {
    ...
  }

  doSomething();
  // [END your_region_tag]
}

main(...process.argv.slice(2));
```

☝️ this will mean that the code between the START/END region tags can be embedded on cloud.google.com, and represents an appropriate body for a cloud function.